### PR TITLE
Avoid caching compilation data and use value equality for SyntaxNodes

### LINKED ISSUE #78242: LibraryImport with custom StringMarshalling type causes the source generator to keep the Compilation object alive
### Description

When using the `StringMarshallingCustomType` property on a `LibraryImport` attribute, the generator caches a symbol object, causing the compilation to be kept alive across runs of the generator.

### Reproduction Steps

I've included a test case (commented out) in this commit [`32f856c` (#77130)](https://github.com/dotnet/runtime/pull/77130/commits/32f856c4007888f6c33c54572319b6fce91c396b)

### Expected behavior

The test should pass.

### Actual behavior

The test fails.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

The problem is that the `LibraryImportData` record type contains a property for the `StringMarshallingCustomType` property from `LibraryImportAttribute` and we put that property into our "data model" type `IncrementalStubGenerationContext`:

https://github.com/dotnet/runtime/blob/2b1f57e08b26879e4fa99670df078b77c848d9e0/src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportGenerator.cs#L30

We need to change how we pass the attribute data between the "compute the model" stage and the "generate code" stage such that the information we need is available, but the symbols are not passed while still having a decently understandable API.

### PR DESCRIPTION
**Pull request by jtschuster**
Fixes https://github.com/dotnet/runtime/issues/78242.

Creates separate types for InteropAttributeData: one that holds compilation data (InteropAttributeCompilationData), and one that holds only the data necessary for the model to create the generated code (InteropAttributeModelData).

This uncovered some issues with record equality in records that use SyntaxNode. For those, we need to override Equals or wrap the SyntaxNode in a type that overrides Equals to use IsEquivalentTo on the SyntaxNode.

There are probably more places where we use SyntaxNode that aren't caught in the current tests.

To make sure every record has the right equality, I wasn't sure if it would be better to override Equals for each of the records, or create a wrapper record struct for each SyntaxNode that implements the equality we want (and implicit casts to and from the SyntaxNode). Then we wouldn't have to explicitly override the equality in each record that has a SyntaxNode. I also overrode both Equals and GetHashCode, but I'm not confident in my GetHashCode implementation. It could also be done with IEquatable.Equals without needing GetHashCode, but that would require implementing the TypeSyntax equality for every type that inherits from ManagedTypeInfo.

### CODE REVIEW

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/InteropAttributeData.cs [hunk 1]
```diff
@@ -24,27 +23,30 @@ public enum InteropAttributeMember
     /// <summary>
     /// Common data for all source-generated-interop trigger attributes
     /// </summary>
-    public record InteropAttributeData
+    public record InteropAttributeCompilationData : InteropAttributeModelData
```
Comment (teo-tsirpanis) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/InteropAttributeData.cs:26]: You should also change the name of the file.

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportData.cs [hunk 2]
```diff
@@ -1,17 +1,27 @@
...
 // The .NET Foundation licenses this file to you under the MIT license.
 
-using System;
-using System.Runtime.InteropServices;
-using Microsoft.CodeAnalysis;
-
 namespace Microsoft.Interop
 {
     /// <summary>
     /// LibraryImportAttribute data
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportData.cs:7]: Doc is out of date.

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportData.cs [hunk 3]
```diff
@@ -1,17 +1,27 @@
...
     /// </summary>
-    internal sealed record LibraryImportData(string ModuleName) : InteropAttributeData
+    internal sealed record LibraryImportCompilationData(string ModuleName) : InteropAttributeCompilationData
+    {
+        public string EntryPoint { get; init; }
+    }
+
+    internal sealed record LibraryImportData(string ModuleName) : InteropAttributeModelData
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportData.cs:14]: Type doc?

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/InteropAttributeData.cs [hunk 4]
```diff
@@ -24,27 +23,30 @@ public enum InteropAttributeMember
...
+    {
+        public INamedTypeSymbol? StringMarshallingCustomType { get; init; }
+    }
+    public record InteropAttributeModelData
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/InteropAttributeData.cs:30]: Doc for type.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/ManualTypeMarshallingHelper.cs [hunk 5]
```diff
@@ -24,6 +24,22 @@ namespace Microsoft.Interop
     public readonly record struct CustomTypeMarshallers(
         ImmutableDictionary<MarshalMode, CustomTypeMarshallerData> Modes)
     {
+        public bool Equals(CustomTypeMarshallers other)
+        {
+            return Modes.Count == other.Modes.Count
+                && !Modes.Except(other.Modes).Any();
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/ManualTypeMarshallingHelper.cs:30]: Took me a second to understand this check. Please drop a comment on this logic.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/ManagedTypeInfo.cs [hunk 6]
```diff
@@ -18,6 +15,19 @@ public abstract record ManagedTypeInfo(string FullTypeName, string DiagnosticFor
...
+        {
+            return other is not null
+                && Syntax.IsEquivalentTo(other.Syntax)
+                && FullTypeName == other.FullTypeName
+                && DiagnosticFormattedName == other.DiagnosticFormattedName;
+        }
+
+        public override int GetHashCode()
+        {
+            return Syntax.GetHashCode() ^ FullTypeName.GetHashCode() ^ DiagnosticFormattedName.GetHashCode();
```
Comment (jkoritzinsky) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/ManagedTypeInfo.cs:28]: I think that the `Syntax.GetHashCode()` call here should be dropped. It doesn't look like `SyntaxNode` overrides `GetHashCode()`, so we'll end up with mismatching hash codes for equal values, which breaks a lot of things.

```suggestion
            return FullTypeName.GetHashCode() ^ DiagnosticFormattedName.GetHashCode();
```

**Review by elinor-fung — APPROVED** [review]
> This uncovered some issues with record equality in records that use SyntaxNode. For those, we need to override Equals or wrap the SyntaxNode in a type that overrides Equals to use IsEquivalentTo on the SyntaxNode.
>
>There are probably more places where we use SyntaxNode that aren't caught in the current tests.

Do you think it is worth filing an issue to do a pass over this?

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/ManagedTypeInfo.cs [hunk 7]
```diff
@@ -18,6 +15,19 @@ public abstract record ManagedTypeInfo(string FullTypeName, string DiagnosticFor
         private TypeSyntax? _syntax;
         public TypeSyntax Syntax => _syntax ??= SyntaxFactory.ParseTypeName(FullTypeName);
 
+        public virtual bool Equals(ManagedTypeInfo other)
```
[the remark this answers is not in the document]
Comment (jtschuster) [reply src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/ManagedTypeInfo.cs:18]: There shouldn't be a null check, thanks for pointing that out!

**Review by jkoritzinsky — APPROVED** [review]

**Comment by jkoritzinsky** [comment]
Only failure is #79874

**Pull request merged**